<a href="https://colab.research.google.com/github/muchammadwr/churn-at-fit.ly/blob/main/notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Analyst Professional Practical Exam Submission

**You can use any tool that you want to do your analysis and create visualizations. Use this template to write up your summary for submission.**

You can use any markdown formatting you wish. If you are not familiar with Markdown, read the [Markdown Guide](https://s3.amazonaws.com/talent-assets.datacamp.com/Markdown+Guide.pdf) before you start.


## 📝 Task List

Your written report should include written text summaries and graphics of the following:
- Data validation:   
  - Describe validation and cleaning steps for every column in the data
- Exploratory Analysis:  
  - Include two different graphics showing single variables only to demonstrate the characteristics of data  
  - Include at least one graphic showing two or more variables to represent the relationship between features
  - Describe your findings
- Definition of a metric for the business to monitor  
  - How should the business use the metric to monitor the business problem
  - Can you estimate initial value(s) for the metric based on the current data
- Final summary including recommendations that the business should undertake

*Start writing report here..*

## Business Understanding

### Background
Over the past two quarters, Fit.ly Tech has observed an increase in customer churn. Customer retention has become increasingly important as the cost of acquiring new customers continues to rise. As a result, leadership needs a clearer understanding of the patterns and potential factors associated with customer churn.

The available data comes from multiple sources, including user activity, customer support, and account information. Since these datasets were developed independently by different teams, they may use different conventions and structures. Therefore, the analysis will combine and assess these sources to identify meaningful patterns in customer engagement, support activity, and plan type.

The findings will be used to help Marketing and Product identify practical, evidence-based actions that can improve customer retention.


### Objective

To identify patterns and potential drivers of customer churn by analyzing customer engagement, support activity, and plan type, and translate these findings into practical, evidence-based actions that Marketing and Product teams can use to improve customer retention.



### Success Criteria

The analysis will be considered successful if it provides:

- A clear understanding of the current churn rate and how it has changed over the past two quarters.
- Identification of customer segments or behaviors associated with higher churn, particularly across engagement, support activity, and plan type.
- Identification of relevant KPIs that leadership can monitor to detect potential churn risks.
- Evidence-based findings that can be translated into practical actions for Marketing and Product.
- A set of measurable retention recommendations that can be evaluated in the following quarter.

A specific churn-reduction target, such as a 15% reduction, should be confirmed with senior leadership before being used as a formal business target.

### Analytical Questions
1. What is the churn trend over the past two quarters?
2. How does engagement relate to churn?
3. How does support activity relate to churn?
4. How does plan type relate to churn?
5. Which customer patterns are associated with higher churn?

## Data Understanding

In [1]:
# Import Library
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
account_info_df = pd.read_csv("https://raw.githubusercontent.com/muchammadwr/churn-at-fit.ly/refs/heads/main/data/da_fitly_account_info.csv")
customer_support_df = pd.read_csv("https://raw.githubusercontent.com/muchammadwr/churn-at-fit.ly/refs/heads/main/data/da_fitly_customer_support.csv")
user_activity_df = pd.read_csv("https://raw.githubusercontent.com/muchammadwr/churn-at-fit.ly/refs/heads/main/data/da_fitly_user_activity.csv")

Fit.ly Tech uses several important data sources for the churn analysis, including account information, user activity, and customer support data. The account information contains customer details such as signup date, state, plan type, plan list price, and churn status. The user activity data records customer actions within the application and will be used to measure customer engagement. The customer support data contains information about support tickets, including ticket time, resolution time, and support channel, which will be used to assess customer support activity. These datasets are maintained independently by different teams and are updated daily, so the data will need to be reviewed and mapped using appropriate customer identifiers before being combined for analysis. The company does not currently plan to purchase additional external data or conduct surveys; therefore, this analysis will focus on the existing internal datasets to identify churn patterns and potential drivers across engagement, support activity, and plan type.

#### `account_info_df`

Information from stakeholder

Account info:
- Signup_date is the day the customer created their account
- State: currently we’re US only - we’ll need a think about our data structure when we
expand internationally, but we’re not thinking about that right now
- Plan types: I’d expect the different types of plan to be captured in the data, but you
might want to check. We have the following tiers: free, basic, pro or enterprise
- Plan list price: this is actual price paid
- Churn status: we mark people as churned once their subscription has been cancelled.

In [5]:
# Show 20 rows dataframe
account_info_df.sample(20)

,customer_id,email,state,plan,plan_list_price,churn_status
146,C10146,user10146@example.com,Florida,Enterprise,148,Y
330,C10330,user10330@example.net,Oklahoma,Enterprise,82,NaN
80,C10080,user10080@example.org,Delaware,Pro,64,NaN
216,C10216,user10216@example.net,South Carolina,Basic,20,NaN
152,C10152,user10152@example.org,North Dakota,Pro,67,Y
106,C10106,user10106@example.org,North Carolina,Enterprise,130,NaN
131,C10131,user10131@example.com,Kentucky,Pro,54,NaN
322,C10322,user10322@example.net,Texas,Pro,34,Y
143,C10143,user10143@example.com,Montana,Enterprise,137,NaN
10,C10010,user10010@example.net,Kansas,Pro,51,NaN


The `account_info_df` dataset provides an initial overview of customer account information. A random sample of 20 rows was examined from a dataset containing **7 features**: `customer_id`, `email`, `state`, `plan`, `plan_list_price`, and `churn_status`.


In [4]:
# Check dataframe information
account_info_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   customer_id      400 non-null    object
 1   email            400 non-null    object
 2   state            400 non-null    object
 3   plan             400 non-null    object
 4   plan_list_price  400 non-null    int64 
 5   churn_status     114 non-null    object
dtypes: int64(1), object(5)
memory usage: 18.9+ KB


Data Information - `account_info_df`

* **`customer_id`**: Unique identifier for each customer. It contains **400 non-null values**, uses the `object` data type, and can be used to identify customers and connect account information with other datasets.

* **`email`**: Customer email address used as contact or customer-identification information. It contains **400 non-null values** and uses the `object` data type.

* **`state`**: U.S. state where the customer is located. It contains **400 non-null values** and uses the `object` data type. This feature can be used to analyze geographic patterns in customer behavior and churn.

* **`plan`**: Customer's subscription tier. It contains **400 non-null values** and uses the `object` data type. Based on the Product Manager's information, the expected plans are **Free, Basic, Pro, and Enterprise**.

* **`plan_list_price`**: The actual price paid by the customer for their subscription. It contains **400 non-null values** and uses the `int64` data type. This feature can be used to analyze customer revenue and compare monetary value across subscription plans.

* **`churn_status`**: Indicates whether a customer's subscription has been cancelled. It contains **114 non-null values** and uses the `object` data type. Based on the Product Manager's definition, `Y` represents a churned customer, while the **286 missing values** likely represent customers who have not been marked as churned.


In [5]:
# Check missing values
account_info_df.isna().sum()

,0
customer_id,0
email,0
state,0
plan,0
plan_list_price,0
churn_status,286


`churn_status` contains 286 missing values, while the other five features are complete. According to the stakeholder, these missing values represent customers who remain active, so they should not be treated as missing or invalid data during churn analysis.


In [7]:
# Check duplicate values
account_info_df.duplicated().sum()

np.int64(0)

There's no duplicate values

In [10]:
# Check value null on churn status
account_info_df['churn_status'].value_counts(dropna=False)

,count
churn_status,
NaN,286
Y,114


The `churn_status` column contains **114 churned customers (`Y`)** and **286 customers who have not churned**, represented by missing values. Based on the stakeholder's explanation, these missing values indicate that the customer's subscription is still active and has not been cancelled.


In [11]:
# Descriptive Statistic
account_info_df.describe(include='all')

,customer_id,email,state,plan,plan_list_price,churn_status
count,400,400,400,400,400.000000,114
unique,400,400,50,4,NaN,1
top,C10399,user10399@example.net,Virginia,Basic,NaN,Y
freq,1,1,16,118,NaN,114
mean,NaN,NaN,NaN,NaN,43.965000,NaN
std,NaN,NaN,NaN,NaN,45.263348,NaN
min,NaN,NaN,NaN,NaN,0.000000,NaN
25%,NaN,NaN,NaN,NaN,0.000000,NaN
50%,NaN,NaN,NaN,NaN,26.000000,NaN
75%,NaN,NaN,NaN,NaN,77.250000,NaN


The `account_info_df` dataset contains **400 customers** across **50 U.S. states** and **4 subscription plans**. Each customer has a unique `customer_id` and email address. The `plan` distribution shows that **Basic** is the most common subscription plan, with **118 customers**. The `plan_list_price` ranges from **0 to 148**, with a median of **26** and an average of approximately **43.97**. The `churn_status` column contains **114 recorded churned customers**, all marked as `Y`, while the remaining 286 records are missing. Therefore, the missing churn values should be investigated before calculating the overall churn rate.


#### `customer_support_df`

Information from stakeholder

Customer support:
- Ticket time is according to when it was received by us - so Pacific time.
- Resolution time is from ticket creation to the ticket being closed
- Channel is the channel it was first received in.
- I don’t know about duplicate or merged tickets - I wouldn’t expect so, but who knows?
- We do occasionally receive GDPR data deletion or “right to be forgotten” requests
through customer support, so we need to make sure we stay compliant with handling
and processing those properly.

In [9]:
# Show 20 rows dataframe
customer_support_df.sample(20)

,ticket_time,user_id,channel,topic,resolution_time_hours,state,comments
36,2025-07-06 11:45:41.850113,10193,phone,other,4.73,0,NaN
347,2025-10-25 20:30:36.641121,10065,phone,billing,16.48,0,NaN
267,2025-06-16 12:45:46.965539,10344,email,technical,17.39,0,NaN
491,2025-08-28 16:42:25.531991,10291,phone,other,25.14,0,NaN
144,2025-10-13 02:00:05.365322,10378,phone,other,1.72,1,NaN
583,2025-08-05 02:16:05.785605,10194,chat,billing,5.45,1,NaN
568,2025-06-15 00:14:50.180863,10091,email,other,17.88,0,Request to remove my data.
594,2025-09-05 07:59:14.459346,10066,email,account,7.27,1,NaN
632,2025-07-23 15:26:56.237034,10108,chat,billing,23.23,1,NaN
59,2025-07-17 04:38:47.724818,10115,chat,billing,20.07,1,NaN


In [10]:
customer_support_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 918 entries, 0 to 917
Data columns (total 7 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   ticket_time            918 non-null    object 
 1   user_id                918 non-null    int64  
 2   channel                918 non-null    object 
 3   topic                  918 non-null    object 
 4   resolution_time_hours  918 non-null    float64
 5   state                  918 non-null    int64  
 6   comments               46 non-null     object 
dtypes: float64(1), int64(2), object(4)
memory usage: 50.3+ KB


Insights

1. **`ticket_time`** — Contains **918 non-null values** with `object` data type, meaning every support ticket has a recorded time, but the column should be converted to `datetime` for time-based analysis.

2. **`user_id`** — Contains **918 non-null values** with `int64` data type, meaning every support ticket is associated with a user and can be used to connect support activity with customer information.

3. **`channel`** — Contains **918 non-null values** with `object` data type, meaning all tickets have a recorded support channel, such as phone, chat, or email.

4. **`topic`** — Contains **918 non-null values** with `object` data type, meaning all tickets have an identified support topic that can be analyzed to understand common customer issues.

5. **`resolution_time_hours`** — Contains **918 non-null values** with `float64` data type, meaning there are no missing resolution times and the variable can be used to analyze support resolution performance.

6. **`state`** — Contains **918 non-null values** with `int64` data type, meaning there are no missing values, but its business meaning should be confirmed because it is represented numerically rather than as a descriptive category.

7. **`comments`** — Contains only **46 non-null values out of 918 records**, meaning approximately **95% of the comments are missing**, which may limit the usefulness of this feature for further analysis.


In [ ]:
customer_support_df.isna().sum()

,0
ticket_time,0
user_id,0
channel,0
topic,0
resolution_time_hours,0
state,0
comments,872


In [ ]:
customer_support_df.duplicated().sum()

np.int64(0)

#### `user_activity_df`

User_activity
- I think this is pretty straightforward, this logs user actions.

In [ ]:
# Show 20 rows dataframe
user_activity_df.sample(20)

,event_time,user_id,event_type
363,2025-09-13 15:38:24.239648,10048,read_article
315,2025-09-15 01:18:48.109488,10312,share_workout
64,2025-08-28 11:28:54.738366,10188,read_article
66,2025-10-09 17:05:47.800934,10120,read_article
226,2025-10-12 03:37:58.337253,10020,read_article
228,2025-11-04 05:02:39.285286,10363,read_article
194,2025-08-03 09:31:23.474648,10075,watch_video
135,2025-09-08 17:37:56.416575,10134,track_workout
121,2025-06-22 15:27:48.578978,10392,read_article
368,2025-10-17 14:10:32.655251,10197,share_workout


In [ ]:
# check dataframe information
user_activity_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 445 entries, 0 to 444
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   event_time  445 non-null    object
 1   user_id     445 non-null    int64 
 2   event_type  445 non-null    object
dtypes: int64(1), object(2)
memory usage: 10.6+ KB


In [ ]:
# Check missing values
user_activity_df.isna().sum()

,0
event_time,0
user_id,0
event_type,0


In [ ]:
# check duplicate values
user_activity_df.duplicated().sum()

np.int64(0)

## Data Preparation

#### `account_info_df`

#### `customer_support_df`

#### `user_activity_df`

## Exploratory Data Analysis (EDA)

## Visualization & Explanatory Analysis

## Conclusion & Recommendation

## ✅ When you have finished...
-  Publish your Workspace using the option on the left
-  Check the published version of your report:
	-  Can you see everything you want us to grade?
    -  Are all the graphics visible?
-  Review the grading rubric. Have you included everything that will be graded?
-  Head back to the [Certification Dashboard](https://app.datacamp.com/certification) to submit your practical exam report and record your presentation